File: [155]minStack.java
/* Runtime: 5 ms, faster than 59.15% of Java online submissions for Min Stack.
 * Memory Usage: 39.7 MB, less than 38.41% of Java online submissions for Min Stack.
 */

class MinStack {
    /** initialize your data structure here. */
    private Stack<Integer> values = new Stack<>();
    private Stack<Integer> min = new Stack<>();

    public MinStack() {
    }

    public void push(int x) {
        if (values.empty() || x <= min.peek()) min.push(x);
        values.push(x);
    }

    public void pop() {
        int x = values.pop();
        if (x == min.peek()) min.pop();
    }

    public int top() {
        return values.peek();
    }

    public int getMin() {
        return min.peek();
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: PythonSubmissions/[155]minStack.py
# Runtime: 72 ms, faster than 33.40% of Python3 online submissions for Min Stack.
# Memory Usage: 16.2 MB, less than 100.00% of Python3 online submissions for Min Stack.
# @stack
# Note: solution trade off speed for space. In Java, integer underflow. 
class MinStack:

    def __init__(self):
        """
        initialize your data structure here.
        """
        self.min = float('inf')
        self.stack = []

    def push(self, x: int) -> None:
        self.stack.append(x - self.min)
        if x < self.min:
            self.min = x

    def pop(self) -> None:
        if not self.stack:
            return
        tmp = self.stack.pop()
        if tmp < 0:
            self.min -= tmp

    def top(self) -> int:
        if not self.stack:
            return
        tmp = self.stack[-1]
        if tmp < 0:
            return self.min
        else:
            return self.min + tmp

    def getMin(self) -> int:
        return self.min


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
